File: 0037-sudoku-solver/0037-sudoku-solver.cpp
#include <cmath>

class Solution {
private:
    vector<vector<char>> ans;
    int boxId(int x, int y){
        return (floor(x/3)*3+y/3);
    }
    
    bool isValid(int x, int y, char val, vector<unordered_set<char>>& boxes, vector<vector<char>>& grid){
        // row wise
       
        int m = grid.size(), n = grid[0].size();
        
        for(int i=0;i<m;i++){
            if(grid[i][y] == val) return false;
        }
        
        for(int j=0;j<n;j++){
            if(grid[x][j] == val) return false;
        }
        
        int idx = boxId(x, y);
        if(boxes[idx].count(val)) return false;
        return true;
    }
    
    void dfs(int k, vector<vector<char>>& board, vector<unordered_set<char>>& boxes, vector<vector<int>>& coords, int& totalDots){
        
//         if(totalDots==0) {
//             if(ans.size()==0) ans=board;
            
//             return;
//         }
        
        if(k>=coords.size()) {
            return;
        };
        
        int x = coords[k][0], y = coords[k][1];
        if(board[x][y] =='.'){
            for(int i=1;i<=9;i++){
                char target = i+'0';
                if(isValid(x, y, target, boxes, board)){
                    int id = boxId(x, y);
                    boxes[id].insert(target);
                    board[x][y] = target;
                    totalDots-=1;
                    dfs(k+1, board, boxes, coords, totalDots);
                    if(totalDots!=0){
                        totalDots+=1;
                        board[x][y] = '.';
                        boxes[id].erase(target);
                    }
                    
                }
            }
               
        }
        
        
    }
    
    
public:
    void solveSudoku(vector<vector<char>>& board) {
        vector<vector<int>> coords;
        int total=0;
        vector<unordered_set<char>> boxes(9);
        
        for(int i=0;i<board.size();i++){
            for(int j=0;j<board[0].size();j++){
                if(board[i][j] == '.'){
                    total+=1;
                    coords.push_back({i, j});
                } else {
                    int id = boxId(i, j);
                    boxes[id].insert(board[i][j]);
                }
            }
        }
        dfs(0, board, boxes, coords, total);
        //board=ans;
        
    }
};

File: 0037-sudoku-solver/0037-sudoku-solver.ts
/**
 Do not return anything, modify board in-place instead.
 */

function boxid(i: number, j:number){
    return Math.floor(i/3)*3 + Math.floor(j/3);
}

function isvalid(val: string, x:number, y:number, board:string[][], boxes: string[][]){
    for(let i=0;i<board.length;i++){
        if(board[i][y]==val) return false;
    }
    
    for(let j=0;j<board[0].length;j++){
        if(board[x][j] == val) return false;
    }
    
    let id = boxid(x, y);
    if(boxes[id].find(z => z==val)!=undefined) return false;
    
    return true;
}


function dfs(idx: number, boxes: string[][], board: string[][], states:number[][]) : boolean{
    if(idx >= states.length) return true;
    
    let x = states[idx][0], y = states[idx][1];
    let id = boxid(x, y);
    
    for(let i=1;i<=9;i++){
        let target = i.toString();
        if(isvalid(target, x, y, board, boxes)){
            board[x][y] = target;
            boxes[id].push(target);
            if(dfs(idx+1, boxes, board, states)) return true;
            boxes[id].pop();
            board[x][y]=".";
        }
    }
    
}

function solveSudoku(board: string[][]): void {

    
    
    let states: number[][] = [];
    let boxes: string[][] = Array.from(Array(9), ()=>new Array());

    
    for(let i=0;i<board.length;i++){
        for(let j=0;j<board[0].length;j++){
            if(board[i][j] == "."){
                let p = [i,j];
                states.push(p);
            } else {
                boxes[boxid(i, j)].push(board[i][j]);
            }
        }
    }
    
    
    
    
    dfs(0, boxes, board, states);
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
};